pkg/reconciler/buildrun/buildrun.go

```diff
@@ -232,6 +242,23 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 			return reconcile.Result{}, nil
 		}
 
+		if buildRun.IsCanceled() && !lastTaskRun.IsCancelled() {
+			ctxlog.Info(ctx, "buildRun marked for cancellation, patching task run", namespace, request.Namespace, name, request.Name)
+			// patch tekton taskrun a la tkn to start tekton's cancelling logic
+			trueParam := true
+			if err := r.patchTaskRun(ctx, lastTaskRun, "replace", "/spec/status", v1beta1.TaskRunSpecStatusCancelled, metav1.PatchOptions{Force: &trueParam}); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
+		succeededCondition := buildRun.Status.GetCondition(buildv1alpha1.Succeeded)
+		if buildRun.IsCanceled() && lastTaskRun.IsCancelled() && (succeededCondition == nil || succeededCondition.Reason != buildv1alpha1.BuildRunSpecStatusCanceled) {
+			if updateErr := resources.UpdateConditionWithFalseStatus(ctx, r.client, buildRun, "the BuildRun is marked canceled.", buildv1alpha1.BuildRunSpecStatusCanceled); updateErr != nil {
+				return reconcile.Result{}, updateErr
+			}
+			return reconcile.Result{}, nil
+		}
```

Comment:
My gut feeling is that this code should not be here, but rather the later code that calls `UpdateBuildRunUsingTaskRunCondition` should handle this case. This will also make sure that a generated service account gets deleted when a TaskRun is canceled (that scenario is maybe worth an integration test). Though, I am not sure what it means about the completion date handling (has a canceled TaskRun has a completionDate ?).

Beside that, I searched for the code that sets the BuildRun status to status=Unknown and reason=BuildRunCanceled? Basically when we are reacting to an update for BuildRunCanceled in the spec and patch the TaskRun, but this is still running.

Reply:
OK circling back to this @SaschaSchwarze0 

First, to answer your question, yes, task run cancelling sets a completion date.  See
- https://github.com/tektoncd/pipeline/blob/7ceaffb1257b5af072f24a0e7b5d837c3ef0a96d/pkg/reconciler/taskrun/taskrun.go#L157 followed by
- https://github.com/tektoncd/pipeline/blob/94bb776234cbc45130a49e27a6f550247553fa20/pkg/reconciler/taskrun/taskrun.go#L568

Now, as it turns out, I originally had this call later in the path here, but ran into headaches when I started adding unit tests, and then rationalized that short circuiting process sooner made sense.

That said, I'll take another pass and look at moving to where the calls to `UpdateBuildRunUsingTaskRunCondition` are and see how things unravel and then how I can subsequently sort them out.

For your "Beside that.." point, I'll circle back to that, after I look into ^^ and I am ready to provide my update to my https://github.com/shipwright-io/build/pull/809#issuecomment-870936103 .... I have an understanding of what is going on with that, but before I dive into it, I want to where I land with ^^ and then level set.

Reply:
In the interim, the full taskrun yaml with what I currently have implementation wise:

```
  apiVersion: tekton.dev/v1beta1
  kind: TaskRun
  metadata:
    annotations:
      pipeline.tekton.dev/release: v0.22.0
    creationTimestamp: "2021-06-30T16:07:55Z"
    generateName: buildpack-nodejs-build-htk8q-
    generation: 2
    labels:
      app.kubernetes.io/managed-by: tekton-pipelines
      build.shipwright.io/generation: "1"
      build.shipwright.io/name: buildpack-nodejs-build
      buildrun.shipwright.io/generation: "1"
      buildrun.shipwright.io/name: buildpack-nodejs-build-htk8q
      clusterbuildstrategy.shipwright.io/generation: "1"
      clusterbuildstrategy.shipwright.io/name: buildpacks-v3
    name: buildpack-nodejs-build-htk8q-88qsq
    namespace: ggmtest
    ownerReferences:
    - apiVersion: shipwright.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: BuildRun
      name: buildpack-nodejs-build-htk8q
      uid: b8993808-520c-4a67-8e6e-1ef154239ee0
    resourceVersion: "172515"
    uid: 29c563ad-b95d-4577-af85-7cf666952df0
  spec:
    params:
    - name: shp-output-image
      value: docker.io/gmontero/sample-nodejs:latest
    - name: shp-source-root
      value: /workspace/source
    - name: CONTEXT_DIR
      value: source-build
    - name: shp-source-context
      value: /workspace/source/source-build
    serviceAccountName: pipeline
    status: TaskRunCancelled
    taskSpec:
      params:
      - default: Dockerfile
        description: Path to the Dockerfile
        name: DOCKERFILE
        type: string
      - default: .
        description: The root of the code
        name: CONTEXT_DIR
        type: string
      - description: The URL of the image that the build produces
        name: shp-output-image
        type: string
      - description: The context directory inside the source directory
        name: shp-source-context
        type: string
      - description: The source directory
        name: shp-source-root
        type: string
      results:
      - description: The digest of the image
        name: shp-image-digest
      - description: The compressed size of the image
        name: shp-image-size
      - description: The commit SHA of the cloned source.
        name: shp-source-default-commit-sha
      steps:
      - args:
        - --url
        - https://github.com/shipwright-io/sample-nodejs
        - --target
        - $(params.shp-source-root)
        - --result-file-commit-sha
        - $(results.shp-source-default-commit-sha.path)
        command:
        - /ko-app/git
        image: docker.io/gmontero/git-77f36d96a091f12d11365ad77da28c66@sha256:9e8bb16e0432a354a10f3caf351093b23028c34013ebcdba41261be7f154216e
        name: source-default
        resources: {}
        securityContext:
          runAsGroup: 1000
          runAsUser: 1000
      - args:
        - -c
        - |
          chown -R "1000:1000" /workspace/source && chown -R "1000:1000" /tekton/home && chown -R "1000:1000" /cache && chown -R "1000:1000" /layers
        command:
        - /bin/bash
        image: docker.io/paketobuildpacks/builder:full
        name: prepare
        resources:
          limits:
            cpu: 500m
            memory: 1Gi
          requests:
            cpu: 250m
            memory: 65Mi
        securityContext:
          capabilities:
            add:
            - CHOWN
          runAsUser: 0
        volumeMounts:
        - mountPath: /cache
          name: cache-dir
        - mountPath: /layers
          name: layers-dir
      - args:
        - -app=/workspace/source/$(inputs.params.CONTEXT_DIR)
        - -cache-dir=/cache
        - -layers=/layers
        - $(params.shp-output-image)
        command:
        - /cnb/lifecycle/creator
        image: docker.io/paketobuildpacks/builder:full
        name: build-and-push
        resources:
          limits:
            cpu: 500m
            memory: 1Gi
          requests:
            cpu: 250m
            memory: 65Mi
        securityContext:
          runAsGroup: 1000
          runAsUser: 1000
        volumeMounts:
        - mountPath: /cache
          name: cache-dir
        - mountPath: /layers
          name: layers-dir
      volumes:
      - name: cache-dir
      - name: layers-dir
      workspaces:
      - name: source
    timeout: 0s
    workspaces:
    - emptyDir: {}
      name: source
  status:
    completionTime: "2021-06-30T16:08:00Z"
    conditions:
    - lastTransitionTime: "2021-06-30T16:08:00Z"
      message: TaskRun "buildpack-nodejs-build-htk8q-88qsq" was cancelled
      reason: TaskRunCancelled
      status: "False"
      type: Succeeded
    podName: buildpack-nodejs-build-htk8q-88qsq-pod-ff8fk
    startTime: "2021-06-30T16:07:55Z"
    steps:
    - container: step-source-default
      imageID: docker.io/gmontero/git-77f36d96a091f12d11365ad77da28c66@sha256:9e8bb16e0432a354a10f3caf351093b23028c34013ebcdba41261be7f154216e
      name: source-default
      terminated:
        exitCode: 1
        finishedAt: "2021-06-30T16:08:00Z"
        reason: TaskRunCancelled
        startedAt: "2021-06-30T16:07:58Z"
    - container: step-prepare
      imageID: docker.io/paketobuildpacks/builder@sha256:e3462130656ff77b7d64f0c1660e3116d70074a0568918f1a0898dc687bf9087
      name: prepare
      terminated:
        exitCode: 1
        finishedAt: "2021-06-30T16:08:00Z"
        reason: TaskRunCancelled
        startedAt: "2021-06-30T16:07:59Z"
    - container: step-build-and-push
      imageID: docker.io/paketobuildpacks/builder@sha256:e3462130656ff77b7d64f0c1660e3116d70074a0568918f1a0898dc687bf9087
      name: build-and-push
      terminated:
        exitCode: 1
        finishedAt: "2021-06-30T16:08:00Z"
        reason: TaskRunCancelled
        startedAt: "2021-06-30T16:07:59Z"
    taskSpec:
      params:
      - default: Dockerfile
        description: Path to the Dockerfile
        name: DOCKERFILE
        type: string
      - default: .
        description: The root of the code
        name: CONTEXT_DIR
        type: string
      - description: The URL of the image that the build produces
        name: shp-output-image
        type: string
      - description: The context directory inside the source directory
        name: shp-source-context
        type: string
      - description: The source directory
        name: shp-source-root
        type: string
      results:
      - description: The digest of the image
        name: shp-image-digest
      - description: The compressed size of the image
        name: shp-image-size
      - description: The commit SHA of the cloned source.
        name: shp-source-default-commit-sha
      steps:
      - args:
        - --url
        - https://github.com/shipwright-io/sample-nodejs
        - --target
        - $(params.shp-source-root)
        - --result-file-commit-sha
        - $(results.shp-source-default-commit-sha.path)
        command:
        - /ko-app/git
        image: docker.io/gmontero/git-77f36d96a091f12d11365ad77da28c66@sha256:9e8bb16e0432a354a10f3caf351093b23028c34013ebcdba41261be7f154216e
        name: source-default
        resources: {}
        securityContext:
          runAsGroup: 1000
          runAsUser: 1000
      - args:
        - -c
        - |
          chown -R "1000:1000" /workspace/source && chown -R "1000:1000" /tekton/home && chown -R "1000:1000" /cache && chown -R "1000:1000" /layers
        command:
        - /bin/bash
        image: docker.io/paketobuildpacks/builder:full
        name: prepare
        resources:
          limits:
            cpu: 500m
            memory: 1Gi
          requests:
            cpu: 250m
            memory: 65Mi
        securityContext:
          capabilities:
            add:
            - CHOWN
          runAsUser: 0
        volumeMounts:
        - mountPath: /cache
          name: cache-dir
        - mountPath: /layers
          name: layers-dir
      - args:
        - -app=/workspace/source/$(inputs.params.CONTEXT_DIR)
        - -cache-dir=/cache
        - -layers=/layers
        - $(params.shp-output-image)
        command:
        - /cnb/lifecycle/creator
        image: docker.io/paketobuildpacks/builder:full
        name: build-and-push
        resources:
          limits:
            cpu: 500m
            memory: 1Gi
          requests:
            cpu: 250m
            memory: 65Mi
        securityContext:
          runAsGroup: 1000
          runAsUser: 1000
        volumeMounts:
        - mountPath: /cache
          name: cache-dir
        - mountPath: /layers
          name: layers-dir
      volumes:
      - name: cache-dir
      - name: layers-dir
      workspaces:
      - name: source
```

Reply:
and the buildrun yaml:

```
  apiVersion: shipwright.io/v1alpha1
  kind: BuildRun
  metadata:
    creationTimestamp: "2021-06-30T16:07:55Z"
    generateName: buildpack-nodejs-build-
    generation: 2
    labels:
      build.shipwright.io/generation: "1"
      build.shipwright.io/name: buildpack-nodejs-build
    name: buildpack-nodejs-build-htk8q
    namespace: ggmtest
    resourceVersion: "172510"
    uid: b8993808-520c-4a67-8e6e-1ef154239ee0
  spec:
    buildRef:
      name: buildpack-nodejs-build
    status: BuildRunCanceled
    timeout: 0s
  status:
    buildSpec:
      output:
        credentials:
          name: push-secret
        image: docker.io/gmontero/sample-nodejs:latest
      source:
        contextDir: source-build
        url: https://github.com/shipwright-io/sample-nodejs
      strategy:
        kind: ClusterBuildStrategy
        name: buildpacks-v3
    completionTime: "2021-06-30T16:08:00Z"
    conditions:
    - lastTransitionTime: "2021-06-30T16:08:00Z"
      message: the BuildRun is marked canceled.
      reason: BuildRunCanceled
      status: "False"
      type: Succeeded
    latestTaskRunRef: buildpack-nodejs-build-htk8q-88qsq
    startTime: "2021-06-30T16:07:55Z"

```

Reply:
> 
> Beside that, I searched for the code that sets the BuildRun status to status=Unknown and reason=BuildRunCanceled? Basically when we are reacting to an update for BuildRunCanceled in the spec and patch the TaskRun, but this is still running.

OK I have the move to `UpdateBuildRunUsingTaskRunCondition` working locally and will push that up in a bit.

On this portion of the comment, I may need some elaboration, though I have a notion of what you may want.  

When I run `kubectl get br -o yaml -w` when running a buildrun, the Succeeded condition remains with status Unknown until the taskrun fails / succeeds / cancels ... i.e. the taskrun reaches a terminal state.

The current code hence does not bother with setting the status of the Succeeded condition when it first sees that the user patched the buildrun with cancel.  Hence it "stays" unknown.  I believe that is because we just use the taskRun condition's status, and it stays unknown.  i.e. https://github.com/shipwright-io/build/blob/main/pkg/reconciler/buildrun/resources/conditions.go#L106

A quick yaml snippet:

```
  conditions:
  - lastTransitionTime: "2021-06-30T16:55:15Z"
    message: Not all Steps in the Task have finished executing
    reason: Running
    status: Unknown
    type: Succeeded

```

Then when it observes the taskrun has been processed, it sets the suceeded condition to false.

Now, I *can* be explicit around making sure this holds.  I'll be including that in my  change, with a comment that we can discuss in this PR about whether to keep it or not.  Among other things, we can provide a message that is more precise doing it this way.

Reply:
^^ change, plus the cancelled build run autogen sa integration test, pushed